### PR TITLE
REV-0026-Create-Map-Component

### PR DESCRIPTION
**Pull request by mhn-opt98**
A Map component contains markers that when hovering shows a specific component(listing , card , etc ...
Screenshot from 2021-12-30 20-02-06
)

### CODE REVIEW

File: src/assets/icons/Group-red.svg [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+            <g id="Group" transform="translate(544.000000, 80.000000)">
+                <g id="Group-151" transform="translate(-3.000000, -172.000000)">
+                    <g id="Group-143" transform="translate(167.000000, 657.000000)">
+                        <path d="M20.1164429,30.8452008 C14.6727611,30.8452008 10.2578066,26.2613917 10.2578066,20.6095105 C10.2578066,14.9576293 14.6727611,10.3738202 20.1164429,10.3738202 C25.5601247,10.3738202 29.9750793,14.9576293 29.9750793,20.6095105 C29.9750793,26.2613917 25.5601247,30.8452008 20.1164429,30.8452008 M20.1164429,0.014510403 C9.16240253,0.014510403 0.280104545,9.23652125 0.280104545,20.6095105 C0.280104545,37.4909824 20.1164429,55.8113846 20.1164429,55.8113846 C20.1164429,55.8113846 39.9527813,37.4909824 39.9527813,20.6095105 C39.9527813,9.23652125 31.0704833,0.014510403 20.1164429,0.014510403" id="Fill-139"></path>
+                    </g>
+                </g>
+            </g>
+        </g>
+    </g>
+</svg>
```
Comment (AseelRousan-op) [inline src/assets/icons/Group-red.svg:15]:
```suggestion
</svg>

```

File: src/components/Map/Marker/index.jsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  onMouseOverHandler: PropTypes.func,
+};
+
+Marker.defaultProps = {
+  data: {
+    title: 'No Title Available',
+  },
+  lat: 0,
+  lng: 0,
+  onMouseOverHandler: () => {},
```
Comment (AseelRousan-op) [inline src/components/Map/Marker/index.jsx:45]: isn't this function required prop here?

File: src/components/Map/index.jsx [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import defaultProps from './map-constants';
+import { GoogleMapContainer } from './map.styles';
+import Marker from './Marker';
+
+/**
+ * Map component.
+ *
+ * @param {React.Component} ComponentOnMap component that will appear on the map when hovering over marker , this component must accept lng , lat , data as props
+ * @param {boolean} isMarkerShown specifiy map to be with or without markers
+ * @param {Array} listings data needed for the Marker
```
Comment (AseelRousan-op) [inline src/components/Map/index.jsx:14]:
```suggestion
 * @param {React.Component} ComponentOnMap Component that will appear on the map when hovering over the marker, this component must accept lng , lat, data as props
 * @param {boolean}         isMarkerShown  Wheather map markers or not
 * @param {Array}           listings       Data needed for the Marker
```

File: src/components/Map/Marker/index.jsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import React from 'react';
+import pinImg from '../../../assets/icons/Group-red.svg';
+
+/**
+ * Map component.
+ *
+ * @param {Object} data data needed for the rendered component
+ * @param {lat} number latitude to specifiy where the element has to be shown on the map
+ * @param {lng} number longitude to specifiy where the element has to be shown on the map
+ * @param {Function} onMouseOverHandler hover event handler
```
Comment (AseelRousan-op) [inline src/components/Map/Marker/index.jsx:11]:
```suggestion
 * @param {Object}  data                Data needed for the rendered component
 * @param {number}  lat                 Latitude to specify where the element has to be shown on the map
 * @param {number}  lag                 Longitude to specify where the element has to be shown on the map
 * @param {Function} onMouseOverHandler Hover event handler
```

**Review by AseelRousan-op — COMMENTED** [review]
Can we update the files structure here to move the `Marker` component out of the `Map` component, we can move them to a separate directory named map in the components directory like this:
components/map/Marker and Map here
or maybe we can move the `Map` to the shared folder since we'll be using this in most of the pages, correct?
cc: @abukhadraop ?

File: src/components/Map/Marker/index.jsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ *
+ * @return {JSX.Element}
+ */
+
```
Comment (Mahmoudzaid2994) [inline src/components/Map/Marker/index.jsx:15]:
```suggestion

```

File: src/components/Map/index.jsx [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+ *
+ * @return {JSX.Element}
+ */
+
```
Comment (Mahmoudzaid2994) [inline src/components/Map/index.jsx:18]:
```suggestion

```

File: src/components/Map/index.jsx [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+function Map({ ComponentOnMap, isMarkerShown, listings }) {
+  const [center, setCenter] = useState();
+  const [data, setData] = useState([]);
+  const [isMarkerHovered, setIsMarkerHovered] = useState(false);
+  const [listingonMap, setListingonMap] = useState({});
+
+  const showListingOnMap = (listingData) => {
```
Comment (Mahmoudzaid2994) [inline src/components/Map/index.jsx:25]: Please add docs for this function

File: src/components/Map/index.jsx [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    setIsMarkerHovered(true);
+    setListingonMap(listingData);
+  };
+
+  function getAverageCoordinate(coordsList, direction) {
```
Comment (Mahmoudzaid2994) [inline src/components/Map/index.jsx:30]: Please add docs for this function

File: src/components/Map/index.jsx [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    const total = coordsList.reduce((acc, current) => acc + current);
+    const coordCenter = total
+      ? total / coordsList.length
+      : defaultProps.center[direction];
+    return coordCenter;
+  }
+
+  const getCenter = () => {
```
Comment (Mahmoudzaid2994) [inline src/components/Map/index.jsx:38]: Please add docs for this function

File: src/components/Map/index.jsx [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+    if (!latDataList || !lngDataList) return defaultProps.center;
+
+    const avgLngCenter = getAverageCoordinate(lngDataList, 'lng');
+    const avgLatCenter = getAverageCoordinate(latDataList, 'lat');
+
+    return { lat: avgLatCenter, lng: avgLngCenter };
+  };
+
+  useEffect(() => {
```
Comment (Mahmoudzaid2994) [inline src/components/Map/index.jsx:51]: Please add docs for this function

**Comment by mhn-opt98** [comment]
> Can we update the files structure here to move the `Marker` component out of the `Map` component, we can move them to a separate directory named map in the components directory like this: components/map/Marker and Map here or maybe we can move the `Map` to the shared folder since we'll be using this in most of the pages, correct? cc: @abukhadraop ?

File: src/components/Map/index.jsx (hunk 8 above)
Comment (AhmadAlShwaiki) [inline src/components/Map/index.jsx:30]: Please commit to using arrow functions.

**Comment by mhn-opt98** [comment]
> Can we update the files structure here to move the `Marker` component out of the `Map` component, we can move them to a separate directory named map in the components directory like this: components/map/Marker and Map here or maybe we can move the `Map` to the shared folder since we'll be using this in most of the pages, correct? cc: @abukhadraop ?

@abukhadraop

File: src/components/Map/index.jsx [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+ * Map component.
+ *
+ * @param {React.Component} ComponentOnMap component that will appear on the map when hovering over marker , this component must accept lng , lat , data as props
+ * @param {boolean} isMarkerShown specifiy map to be with or without markers
+ * @param {Array} listings data needed for the Marker
+ *
+ * @return {JSX.Element}
+ */
+
+function Map({ ComponentOnMap, isMarkerShown, listings }) {
```
Comment (AhmadAlShwaiki) [inline src/components/Map/index.jsx:19]: Please commit to camelCase for variables.
```
function Map({ componentOnMap, isMarkerShown, listings }) {
```

**Review by AhmadAlShwaiki — CHANGES_REQUESTED** [review]
Suggestion for naming and linting best practices

File: src/pages/Home.jsx [hunk 12]
```diff
@@ -1,13 +1,48 @@
...
+DetailedCard.defaultProps = {
+  data: {},
+};
+
+const maxLong = -74.871826;
+const minLong = -74.9;
+const minLat = 39.833851;
+const maxLat = 39.84;
+
+const randomIntFromInterval = (min, max) => Math.random() * (max - min) + min;
```
Comment (AhmadAlShwaiki) [inline src/pages/Home.jsx:36]: Please use function documentation

File: src/pages/Home.jsx [hunk 13]
```diff
@@ -1,13 +1,48 @@
...
+};
+
+DetailedCard.defaultProps = {
+  data: {},
+};
+
+const maxLong = -74.871826;
+const minLong = -74.9;
+const minLat = 39.833851;
+const maxLat = 39.84;
```
Comment (AhmadAlShwaiki) [inline src/pages/Home.jsx:34]: I can't see why do we need these hardcoded values.
If necessary you can add them in a separate file in one object 
for example:
```
const MAXIMUM_COORDINATES : {
```

File: src/components/shared/Map/index.jsx [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+      setCenter(getCenter());
+    };
+
+    setupMap();
+  }, [listings]);
+
+  return (
+    <GoogleMapContainer>
+      <GoogleMapReact
+        bootstrapURLKeys={defaultProps.apiKey}
```
Comment (AhmadAlShwaiki) [inline src/components/shared/Map/index.jsx:81]: Can we use object destruct for the defaultProps? 
const { apiKey } = defaultProps;

File: src/components/shared/Marker/index.jsx [hunk 15]
```diff
@@ -0,0 +1,46 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import pinImg from '../../../assets/icons/Group-red.svg';
```
Comment (AhmadAlShwaiki) [inline src/components/shared/Marker/index.jsx:3]:
```suggestion
import pinImg from 'assets/icons/Group-red.svg';
```

**Review by Mahmoudzaid2994 — APPROVED** [review]

**Comment by AseelRousan-op** [comment]
> > Can we update the files structure here to move the `Marker` component out of the `Map` component, we can move them to a separate directory named map in the components directory like this: components/map/Marker and Map here or maybe we can move the `Map` to the shared folder since we'll be using this in most of the pages, correct? cc: @abukhadraop ?
> 
> @abukhadraop

since we moved the Map component to the shared directory here, I'm gonna approve it in advance 👍

**Review by AseelRousan-op — APPROVED** [review]

**Pull request merged**